extmod/nimble/nimble/nimble_npl_os.c

```diff
@@ -493,23 +501,18 @@ void ble_npl_time_delay(ble_npl_time_t ticks) {
 // CRITICAL
 
 // This is used anywhere NimBLE modifies global data structures.
-// We need to protect between:
-//  - A MicroPython VM thread.
-//  - The NimBLE "task" (e.g. PENDSV on STM32, pthread on Unix).
-// On STM32, by disabling PENDSV, we ensure that either:
-//  - If we're in the NimBLE task, we're exclusive anyway.
-//  - If we're in a VM thread, we can't be interrupted by the NimBLE task, or switched to another thread.
-// On Unix, there's a global mutex.
 
-// TODO: Both ports currently use MICROPY_PY_BLUETOOTH_ENTER in their implementation,
-// maybe this doesn't need to be port-specific?
+// Currently all NimBLE code is invoked by the scheduler so there should be no
+// races. However, in the future we may wish to make HCI UART processing
+// happen asynchronously (e.g. on RX IRQ), so these provide a safe
+// implementation.
 
 uint32_t ble_npl_hw_enter_critical(void) {
     DEBUG_CRIT_printf("ble_npl_hw_enter_critical()\n");
-    return mp_bluetooth_nimble_hci_uart_enter_critical();
+    return MICROPY_BEGIN_ATOMIC_SECTION();
```

Comment:
On STM32 when sync events are enabled, this can actually be a no-op as all NimBLE code is now running in the scheduler. I've made it use BLUETOOTH_ENTER/EXIT, made that a no-op on STM32-with-sync-events and updated the comments.